In [347]:
import pandas as pd

โหลดข้อมูลมาจากไฟล์ student.csv เข้ามาไว้ในตัวแปร df_master

In [348]:
df_master = pd.read_csv("student.csv")

ดูข้อมูลจากไฟล์ csv

In [349]:
df_master.head(10)

รหัสนักศึกษา      ชื่อ   อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0         64001     สมชาย   20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1         64002    สมหญิง   21.0  หญิง           วิทย์  3.45       165.0   
2         64003       NaN    NaN  Male         Science  3.67       175.0   
3         64004  สมศักดิ์   22.0     ช  คณะวิทยาศาสตร์ -1.00       168.0   
4         64005      สมใจ   19.0     ญ           วิศวะ  3.89       160.0   
5         64006     กานต์  150.0   ชาย     Engineering  3.12       172.0   
6         64007     กิตติ   21.0  หญิง        วิศวกรรม  4.50       158.0   
7         64008      จิรา   20.0     F      ศิลปศาสตร์  2.98       180.0   
8         64009    ชนาธิป   23.0     M            Arts  3.34       163.0   
9         64010      ดารา   21.0  หญิง           ศิลป์  3.56       167.0   

   น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0        65.0                     15.0      2024-01-23  
1        58.0                     20.0      2024-02-05  
2         NaN                     25.0      2024-01-13  
3        70.0                     10.0      2023-12-30  
4        52.0                     30.0      2024-01-27  
5        75.0                     12.0      2023-12-23  
6        48.0                     35.0      2024-01-05  
7        82.0                     18.0      2024-01-23  
8        55.0                     22.0      2024-02-11  
9        60.0                     28.0      2024-01-03

In [350]:
import numpy as np
from datetime import datetime, timedelta
import warnings
warnings.filterwarnings('ignore')

In [351]:
np.random.seed(42)

สร้าง คอลัมน์วันที่ลงทะเบียนแบบสุ่ม ให้กับตารางข้อมูล df_master ที่สุ่ม ขึ้นมา 20 วัน โดย:สุ่มจำนวนเต็มระหว่าง -30 ถึง +29 และ ทำทั้งหมดนี้ 20 รอบ แล้วสร้างเป็นลิสต์ของวันที่ จากนั้นเอาผลลัพธ์ไปใส่ในคอลัมน์ใหม่ของ DataFrame ชื่อว่า "วันที่ลงทะเบียน"





In [352]:
base_date = datetime(2024, 1, 15)
df_master['วันที่ลงทะเบียน'] = [base_date + timedelta(days=np.random.randint(-30, 30))
                                for _ in range(20)]

ดูข้อมูลล่าสุดว่ามีอะไรบ้างจากตัวแปร df_master (เรียกดู 10 อันดับแรก)

In [353]:
df_master.head(10)

รหัสนักศึกษา      ชื่อ   อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0         64001     สมชาย   20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1         64002    สมหญิง   21.0  หญิง           วิทย์  3.45       165.0   
2         64003       NaN    NaN  Male         Science  3.67       175.0   
3         64004  สมศักดิ์   22.0     ช  คณะวิทยาศาสตร์ -1.00       168.0   
4         64005      สมใจ   19.0     ญ           วิศวะ  3.89       160.0   
5         64006     กานต์  150.0   ชาย     Engineering  3.12       172.0   
6         64007     กิตติ   21.0  หญิง        วิศวกรรม  4.50       158.0   
7         64008      จิรา   20.0     F      ศิลปศาสตร์  2.98       180.0   
8         64009    ชนาธิป   23.0     M            Arts  3.34       163.0   
9         64010      ดารา   21.0  หญิง           ศิลป์  3.56       167.0   

   น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0        65.0                     15.0      2024-01-23  
1        58.0                     20.0      2024-02-05  
2         NaN                     25.0      2024-01-13  
3        70.0                     10.0      2023-12-30  
4        52.0                     30.0      2024-01-27  
5        75.0                     12.0      2023-12-23  
6        48.0                     35.0      2024-01-05  
7        82.0                     18.0      2024-01-23  
8        55.0                     22.0      2024-02-11  
9        60.0                     28.0      2024-01-03


- ตรวจสอบค่า missing value (ดูว่าค่าไหนเป็น NAN) และทำการนับว่ามีค่าที่เป็นค่า missing value อยู่กี่ค่า
- ตรวจสอบว่ามีอายุผิดปกติไหม ( มากกว่า 100 ) โดยเช็คว่ามีอย่างน้อย 1 แถวที่ตรงเงื่อนไขไหม ถ้ามีให้แสดงค่าที่ผิดปกติ ถ้าไม่มีให้แสดงคำว่า "ไม่มี"
- ตรวจสอบ GPA ผิดปกติ ( ค่าที่ไม่อยู่ในช่วง 0-4 ) โดยตรวจหา GPA ที่ไม่สมเหตุสมผล โดยใช้ any() เหมือนด้านบน แสดงค่าที่ผิดปกติ หรือบอกว่า "ไม่มี"
- ตรวจสอบจำนวนรูปแบบการเขียนชื่อ "คณะ" ว่ามีการเขียนชื่อคณะหลากหลายรูปแบบเกินไปไหม
- ตรวจสอบรูปแบบการเขียน "เพศ" ว่ามีการเขียนหลายรูปแบบไหม


In [354]:
df_master.isnull()

รหัสนักศึกษา   ชื่อ   อายุ    เพศ    คณะ    GPA  ส่วนสูง_cm  น้ำหนัก_kg  \
0          False  False  False  False  False  False       False       False   
1          False  False  False  False  False  False       False       False   
2          False   True   True  False  False  False       False        True   
3          False  False  False  False  False  False       False       False   
4          False  False  False  False  False  False       False       False   
5          False  False  False  False  False  False       False       False   
6          False  False  False  False  False  False       False       False   
7          False  False  False  False  False  False       False       False   
8          False  False  False  False  False  False       False       False   
9          False  False  False  False  False  False       False       False   
10         False  False   True   True   True  False        True       False   
11         False  False  False  False  False   True       False        True   
12         False  False  False  False  False  False       False       False   
13         False  False  False  False  False  False       False       False   
14         False  False  False  False  False  False       False       False   
15         False  False  False  False  False  False       False       False   
16         False  False  False  False  False  False       False       False   
17         False  False  False  False   True  False       False       False   
18         False  False   True  False  False   True       False       False   
19         False  False  False   True  False  False       False       False   

    ชั่วโมงเรียน_ต่อสัปดาห์  วันที่ลงทะเบียน  
0                     False            False  
1                     False            False  
2                     False            False  
3                     False            False  
4                     False            False  
5                     False            False  
6                     False            False  
7                     False            False  
8                     False            False  
9                     False            False  
10                     True            False  
11                    False            False  
12                    False            False  
13                    False            False  
14                    False            False  
15                    False            False  
16                    False            False  
17                    False            False  
18                     True            False  
19                    False            False

In [355]:
df_master.isnull().sum().sum()

np.int64(15)

In [356]:
df_master[df_master['อายุ'] > 100]['อายุ'].values if any(df_master['อายุ'] > 100) else 'ไม่มี'

array([150.])

In [357]:
df_master[(df_master['GPA'] < 0) | (df_master['GPA'] > 4)]['GPA'].values if any((df_master['GPA'] < 0) | (df_master['GPA'] > 4)) else 'ไม่มี'

array([-1. ,  4.5])

In [358]:
df_master['คณะ'].nunique()

14

In [359]:
df_master['เพศ'].nunique()

8

In [360]:
print(" ข้อมูลหลัก (Master Dataset) ที่จะใช้ตลอดบทเรียน")
print("\n")
print(df_master)
print("\n สรุปปัญหาที่พบในข้อมูล:")
print(f"- ค่าหายทั้งหมด: {df_master.isnull().sum().sum()} ค่า")
print(f"- อายุผิดปกติ: {df_master[df_master['อายุ'] > 100]['อายุ'].values if any(df_master['อายุ'] > 100) else 'ไม่มี'}")
print(f"- GPA ผิดปกติ: {df_master[(df_master['GPA'] < 0) | (df_master['GPA'] > 4)]['GPA'].values if any((df_master['GPA'] < 0) | (df_master['GPA'] > 4)) else 'ไม่มี'}")
print(f"- รูปแบบการเขียนคณะ: {df_master['คณะ'].nunique()} แบบ")
print(f"- รูปแบบการเขียนเพศ: {df_master['เพศ'].nunique()} แบบ")

 ข้อมูลหลัก (Master Dataset) ที่จะใช้ตลอดบทเรียน


    รหัสนักศึกษา      ชื่อ   อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0    หญิง           วิทย์  3.45       165.0   
2          64003       NaN    NaN    Male         Science  3.67       175.0   
3          64004  สมศักดิ์   22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ   19.0       ญ           วิศวะ  3.89       160.0   
5          64006     กานต์  150.0     ชาย     Engineering  3.12       172.0   
6          64007     กิตติ   21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา   20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0       M            Arts  3.34       163.0   
9          64010      ดารา   21.0    หญิง           ศิลป์  3.56       167.0   
10         64011      ธีระ    NaN     NaN             NaN  3.78         NaN   
1

ทำการสำรองข้อมูล กันข้อมูลมีปัญหา

In [361]:
df_original = df_master.copy()

ดูภาพรวมของข้อมูล เช่น ขนาด, คอลัมน์, ค่าว่าง, ประเภทข้อมูล

In [362]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   รหัสนักศึกษา             20 non-null     int64         
 1   ชื่อ                     19 non-null     object        
 2   อายุ                     17 non-null     float64       
 3   เพศ                      18 non-null     object        
 4   คณะ                      18 non-null     object        
 5   GPA                      18 non-null     float64       
 6   ส่วนสูง_cm               19 non-null     float64       
 7   น้ำหนัก_kg               18 non-null     float64       
 8   ชั่วโมงเรียน_ต่อสัปดาห์  18 non-null     float64       
 9   วันที่ลงทะเบียน          20 non-null     datetime64[ns]
dtypes: datetime64[ns](1), float64(5), int64(1), object(3)
memory usage: 1.7+ KB


ดูค่าสถิติเบื้องต้นของคอลัมน์ตัวเลขใน DataFrame

In [363]:
df_master.describe()

รหัสนักศึกษา        อายุ        GPA  ส่วนสูง_cm  น้ำหนัก_kg  \
count      20.00000   17.000000  18.000000   19.000000   18.000000   
mean    64010.50000   28.470588   3.246111  167.684211   63.611111   
min     64001.00000   19.000000  -1.000000  158.000000   48.000000   
25%     64005.75000   20.000000   3.215000  163.500000   55.250000   
50%     64010.50000   21.000000   3.450000  168.000000   63.000000   
75%     64015.25000   22.000000   3.670000  171.500000   71.500000   
max     64020.00000  150.000000   4.500000  180.000000   82.000000   
std         5.91608   31.337114   1.123326    5.706260   10.279270   

       ชั่วโมงเรียน_ต่อสัปดาห์      วันที่ลงทะเบียน  
count                18.000000                   20  
mean                 21.388889  2024-01-12 12:00:00  
min                  10.000000  2023-12-18 00:00:00  
25%                  16.500000  2024-01-02 00:00:00  
50%                  21.500000  2024-01-08 00:00:00  
75%                  25.750000  2024-01-23 06:00:00  
max                  35.000000  2024-02-11 00:00:00  
std                   6.598920                  NaN

ตรวจสอบค่าหายในแต่ละคอลัมน์

In [364]:
missing_counts = df_master.isnull().sum()

In [365]:
missing_counts

รหัสนักศึกษา               0
ชื่อ                       1
อายุ                       3
เพศ                        2
คณะ                        2
GPA                        2
ส่วนสูง_cm                 1
น้ำหนัก_kg                 2
ชั่วโมงเรียน_ต่อสัปดาห์    2
วันที่ลงทะเบียน            0
dtype: int64

แสดงผลค่าว่าง (NaN) ในแต่ละคอลัมน์ ของ DataFrame df_master โดยแสดงทั้งจำนวนและเปอร์เซ็นต์ของข้อมูลที่หายไปในแต่ละคอลัมน์แบบสวยงาม
ด้วยการวนลูปใช้สำหรับ แสดงคอลัมน์ที่มีข้อมูลหาย (NaN) พร้อมทั้งแสดง จำนวนค่าว่างและเปอร์เซ็นต์ของค่าว่างเทียบกับจำนวนข้อมูลทั้งหมด

In [366]:
print("\n ค่าหายในแต่ละคอลัมน์:")
for col, count in missing_counts[missing_counts > 0].items():
    percentage = (count / len(df_master)) * 100
    print(f"   - {col}: {count} ค่า ({percentage:.1f}%)")


 ค่าหายในแต่ละคอลัมน์:
   - ชื่อ: 1 ค่า (5.0%)
   - อายุ: 3 ค่า (15.0%)
   - เพศ: 2 ค่า (10.0%)
   - คณะ: 2 ค่า (10.0%)
   - GPA: 2 ค่า (10.0%)
   - ส่วนสูง_cm: 1 ค่า (5.0%)
   - น้ำหนัก_kg: 2 ค่า (10.0%)
   - ชั่วโมงเรียน_ต่อสัปดาห์: 2 ค่า (10.0%)


แสดงข้อมูลปัจจุบัน

In [367]:
df_original

รหัสนักศึกษา      ชื่อ   อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0    หญิง           วิทย์  3.45       165.0   
2          64003       NaN    NaN    Male         Science  3.67       175.0   
3          64004  สมศักดิ์   22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ   19.0       ญ           วิศวะ  3.89       160.0   
5          64006     กานต์  150.0     ชาย     Engineering  3.12       172.0   
6          64007     กิตติ   21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา   20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0       M            Arts  3.34       163.0   
9          64010      ดารา   21.0    หญิง           ศิลป์  3.56       167.0   
10         64011      ธีระ    NaN     NaN             NaN  3.78         NaN   
11         64012       นภา   22.0     ชาย           วิทย์   NaN       171.0   
12         64013     บุญมี   20.0    หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0     ชาย    Liberal Arts  3.45       164.0   
14         64015     พิมพ์   22.0  Female             Sci  3.67       162.0   
15         64016      ภูมิ   19.0       ช     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0       ญ  วิศวกรรมศาสตร์  3.01       166.0   
17         64018      ยุพา   21.0    หญิง             NaN  3.23       161.0   
18         64019     รัตนา    NaN     ชาย      ศิลปศาสตร์   NaN       174.0   
19         64020     ลัดดา   22.0     NaN         Science  3.45       168.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
2          NaN                     25.0      2024-01-13  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
10        68.0                      NaN      2024-01-07  
11         NaN                     16.0      2023-12-26  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24  
17        50.0                     27.0      2024-01-08  
18        76.0                      NaN      2023-12-18  
19        64.0                     23.0      2024-01-06

เราจะมาลองอีกวิธีคือ วิธีการเติมค่าด้วยค่าเฉลี่ย ในส่วนนี้จะทำในคอลัมน์ "อายุ" ด้วยการ เติมค่าที่หาย (NaN) ในคอลัมน์ "อายุ" ด้วย "ค่าเฉลี่ยของอายุทั้งหมด" ซึ่งวิธีนี้ใช้ได้กับค่าที่เป็น Nan เท่านั้น



In [368]:
mean_age = df_master['อายุ'].mean()
df_master['อายุ'].fillna(mean_age, inplace=True)

In [369]:
print(f"\nค่าเฉลี่ยอายุ - {mean_age:.1f} ปี")
print("\n")
df_master


ค่าเฉลี่ยอายุ - 28.5 ปี




รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN             NaN  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง             NaN  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0        170.0        65.0                     15.0      2024-01-23  
1        165.0        58.0                     20.0      2024-02-05  
2        175.0         NaN                     25.0      2024-01-13  
3        168.0        70.0                     10.0      2023-12-30  
4        160.0        52.0                     30.0      2024-01-27  
5        172.0        75.0                     12.0      2023-12-23  
6        158.0        48.0                     35.0      2024-01-05  
7        180.0        82.0                     18.0      2024-01-23  
8        163.0        55.0                     22.0      2024-02-11  
9        167.0        60.0                     28.0      2024-01-03  
10         NaN        68.0                      NaN      2024-01-07  
11       171.0         NaN                     16.0      2023-12-26  
12       169.0        72.0                     24.0      2023-12-26  
13       164.0        54.0                     19.0      2024-01-08  
14       162.0        56.0                     26.0      2024-02-06  
15       173.0        78.0                     14.0      2024-01-20  
16       166.0        62.0                     21.0      2024-01-24  
17       161.0        50.0                     27.0      2024-01-08  
18       174.0        76.0                      NaN      2023-12-18  
19       168.0        64.0                     23.0      2024-01-06

ต่อมาเราจะมาแก้ไขในส่วนของ "คณะ" เราจะทำการรับค่า name (ซึ่งคือชื่อคณะหรือคอลัมน์ที่ต้องการจัดการ)

ตรวจสอบว่า name ตรงกับคำที่หลากหลาย (รูปแบบเขียนต่างกัน) ที่แสดงถึงคณะเดียวกันหรือไม่

ถ้าใช่ จะ แทนที่ให้เป็นชื่อคณะที่มาตรฐานเดียวกัน เช่น "วิทยาศาสตร์" หรือ "ศิลปศาสตร์"

ถ้าไม่ตรงกับกลุ่มใดเลย จะคืนค่าเดิม (name) กลับไป



In [370]:
def standardize_faculty(name):
    if name in ['วิทยาศาสตร์', 'วิทย์', 'Science', 'คณะวิทยาศาสตร์','Sci']:
        return 'วิทยาศาสตร์'
    elif name in ['ศิลปศาสตร์', 'ศิลป์', 'Arts','คณะศิลปศาสตร์','Liberal Arts']:
        return 'ศิลปศาสตร์'
    elif name in ['วิศวะ', 'Engineering', 'วิศวกรรม', 'วิศวกรรมศาสตร์','คณะวิศวกรรมศาสตร์','วิศวฯ']:
        return 'วิศวกรรมศาสตร์'
    else:
        return name

In [371]:
df_master['คณะ'] = df_master['คณะ'].apply(standardize_faculty)

เรียกดูข้อมูลหลังจากแก้ไขค่าต่างๆในคอลัมน์ "คณะ" ให้ตรงกันแล้ว

In [372]:
df_master

รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง     วิทยาศาสตร์  3.45   
2          64003       NaN   28.470588    Male     วิทยาศาสตร์  3.67   
3          64004  สมศักดิ์   22.000000       ช     วิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ  วิศวกรรมศาสตร์  3.89   
5          64006     กานต์  150.000000     ชาย  วิศวกรรมศาสตร์  3.12   
6          64007     กิตติ   21.000000    หญิง  วิศวกรรมศาสตร์  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M      ศิลปศาสตร์  3.34   
9          64010      ดารา   21.000000    หญิง      ศิลปศาสตร์  3.56   
10         64011      ธีระ   28.470588     NaN             NaN  3.78   
11         64012       นภา   22.000000     ชาย     วิทยาศาสตร์   NaN   
12         64013     บุญมี   20.000000    หญิง  วิศวกรรมศาสตร์  3.23   
13         64014     ปรีชา   21.000000     ชาย      ศิลปศาสตร์  3.45   
14         64015     พิมพ์   22.000000  Female     วิทยาศาสตร์  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง             NaN  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN     วิทยาศาสตร์  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0        170.0        65.0                     15.0      2024-01-23  
1        165.0        58.0                     20.0      2024-02-05  
2        175.0         NaN                     25.0      2024-01-13  
3        168.0        70.0                     10.0      2023-12-30  
4        160.0        52.0                     30.0      2024-01-27  
5        172.0        75.0                     12.0      2023-12-23  
6        158.0        48.0                     35.0      2024-01-05  
7        180.0        82.0                     18.0      2024-01-23  
8        163.0        55.0                     22.0      2024-02-11  
9        167.0        60.0                     28.0      2024-01-03  
10         NaN        68.0                      NaN      2024-01-07  
11       171.0         NaN                     16.0      2023-12-26  
12       169.0        72.0                     24.0      2023-12-26  
13       164.0        54.0                     19.0      2024-01-08  
14       162.0        56.0                     26.0      2024-02-06  
15       173.0        78.0                     14.0      2024-01-20  
16       166.0        62.0                     21.0      2024-01-24  
17       161.0        50.0                     27.0      2024-01-08  
18       174.0        76.0                      NaN      2023-12-18  
19       168.0        64.0                     23.0      2024-01-06

ต่อมาก็เช่นเดิม เราจะมาแก้ไขในส่วนของ "เพศ" เนื่องจากคอลัมน์นี้มีรูปแบบเหมือนกับคอลัมน์ "คณะ" เลยเราจึงแก้ไขค่าของมันด้วยวิธีเดิมคือ เราจะทำการรับค่า name (ซึ่งคือชื่อคณะหรือคอลัมน์ที่ต้องการจัดการ)

ตรวจสอบว่า name ตรงกับคำที่หลากหลาย (รูปแบบเขียนต่างกัน) ที่แสดงถึงคณะเดียวกันหรือไม่

ถ้าใช่ จะ แทนที่ให้เป็นชื่อคณะที่มาตรฐานเดียวกัน เช่น "ชาย" หรือ "หญิง"

ถ้าไม่ตรงกับกลุ่มใดเลย จะคืนค่าเดิม (name) กลับไป

In [373]:
def standardize_gender(gender):
    if gender in ['ชาย', 'M', 'Male', 'ช']:
        return 'ชาย'
    elif gender in ['หญิง', 'F', 'Female', 'ญ']:
        return 'หญิง'
    else:
        return gender

In [374]:
df_master['เพศ'] = df_master['เพศ'].apply(standardize_gender)

In [375]:
df_master

รหัสนักศึกษา      ชื่อ        อายุ   เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000   ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000  หญิง     วิทยาศาสตร์  3.45   
2          64003       NaN   28.470588   ชาย     วิทยาศาสตร์  3.67   
3          64004  สมศักดิ์   22.000000   ชาย     วิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000  หญิง  วิศวกรรมศาสตร์  3.89   
5          64006     กานต์  150.000000   ชาย  วิศวกรรมศาสตร์  3.12   
6          64007     กิตติ   21.000000  หญิง  วิศวกรรมศาสตร์  4.50   
7          64008      จิรา   20.000000  หญิง      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000   ชาย      ศิลปศาสตร์  3.34   
9          64010      ดารา   21.000000  หญิง      ศิลปศาสตร์  3.56   
10         64011      ธีระ   28.470588   NaN             NaN  3.78   
11         64012       นภา   22.000000   ชาย     วิทยาศาสตร์   NaN   
12         64013     บุญมี   20.000000  หญิง  วิศวกรรมศาสตร์  3.23   
13         64014     ปรีชา   21.000000   ชาย      ศิลปศาสตร์  3.45   
14         64015     พิมพ์   22.000000  หญิง     วิทยาศาสตร์  3.67   
15         64016      ภูมิ   19.000000   ชาย     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000  หญิง  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000  หญิง             NaN  3.23   
18         64019     รัตนา   28.470588   ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000   NaN     วิทยาศาสตร์  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0        170.0        65.0                     15.0      2024-01-23  
1        165.0        58.0                     20.0      2024-02-05  
2        175.0         NaN                     25.0      2024-01-13  
3        168.0        70.0                     10.0      2023-12-30  
4        160.0        52.0                     30.0      2024-01-27  
5        172.0        75.0                     12.0      2023-12-23  
6        158.0        48.0                     35.0      2024-01-05  
7        180.0        82.0                     18.0      2024-01-23  
8        163.0        55.0                     22.0      2024-02-11  
9        167.0        60.0                     28.0      2024-01-03  
10         NaN        68.0                      NaN      2024-01-07  
11       171.0         NaN                     16.0      2023-12-26  
12       169.0        72.0                     24.0      2023-12-26  
13       164.0        54.0                     19.0      2024-01-08  
14       162.0        56.0                     26.0      2024-02-06  
15       173.0        78.0                     14.0      2024-01-20  
16       166.0        62.0                     21.0      2024-01-24  
17       161.0        50.0                     27.0      2024-01-08  
18       174.0        76.0                      NaN      2023-12-18  
19       168.0        64.0                     23.0      2024-01-06

ต่อมาเราจะเห็นว่าข้อมูลในคอลันม์ "อายุ" นั้นมีค่าที่แปลกๆคือ 150 ซึ่งหากมองตามความเป็นจริงแล้ว คนเราไม่น่ามีใครอายุ 150 ปีแล้วเป็นนักศึกษาอยู่ ซึ่งการแก้ไขด้วยการการเติมค่าด้วยค่าเฉลี่ย นั้นยังดูไม่ค่อยเหมาะสักเท่าไรเราจึงจะแก้ไขใหม่ด้วยการ การกรองค่าผิดปกติแทน


"การกรองค่าผิดปกติ"
เราจะเริ่มด้วยดูค่าเฉลี่ยอายุที่ไม่มีค่าผิดปกติ (อายุเกิน 100)

In [376]:
print(f"\nค่าเฉลี่ยอายุ (รวมค่าผิดปกติ) - {df_master['อายุ'].mean():.2f}")


ค่าเฉลี่ยอายุ (รวมค่าผิดปกติ) - 28.47


ต่อมาเราจะใช้ IQR (Interquartile Range Method) ในการตรวจจับ Outlier (ค่าผิดปกติ) ในข้อมูล โดยเฉพาะข้อมูลตัวเลข เช่น "อายุ" หรือ "รายได้"

เราจะกำหนดขอบเขตซึ่ง Q1 → ควอไทล์ที่ 1 (25% แรกของข้อมูล)

Q3 → ควอไทล์ที่ 3 (75% แรกของข้อมูล) ตัวเลข 1.5 เป็นค่ามาตรฐานที่นิยมใช้ในการหาค่าผิดปกติ

IQR = Q3 - Q1 → ช่วงระหว่างควอไทล์ (Interquartile Range) ใช้วัดว่าข้อมูลช่วงกลางกระจายมากแค่ไหน

In [377]:
Q1 = df['อายุ'].quantile(0.25)
Q3 = df['อายุ'].quantile(0.75)
IQR = Q3 - Q1

lower_bound และ upper_bound คือ ขอบเขตล่างและบนสำหรับตัดสินว่าค่าใดเป็นค่าผิดปกติ

In [378]:
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

ต่อมาเราจะกรองเฉพาะแถวที่มีอายุอยู่ในช่วงที่กำหนด

& คือตัวดำเนินการ AND ใน pandas (ไม่ใช่ and)

วิธีนี้จะช่วยตัดค่าที่ออกนอกกลุ่มข้อมูลส่วนใหญ่มากเกินไป ซึ่งก็คือหากเจอค่า "อายุ" ที่เป็น outlier ตามเกณฑ์ IQR จึงถูกลบออกนั่นเอง

In [379]:
df_master = df_master[(df_master['อายุ'] >= lower_bound) & (df_master['อายุ'] <= upper_bound)]

เรียกดูข้อมูลล่าสุดหลังจากทำการกรองค่าผิดปกติ

In [380]:
df_master

รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  4.50       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
11         64012       นภา  22.0   ชาย     วิทยาศาสตร์   NaN       171.0   
12         64013     บุญมี  20.0  หญิง  วิศวกรรมศาสตร์  3.23       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยาศาสตร์  3.67       162.0   
15         64016      ภูมิ  19.0   ชาย     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี  20.0  หญิง  วิศวกรรมศาสตร์  3.01       166.0   
17         64018      ยุพา  21.0  หญิง             NaN  3.23       161.0   
19         64020     ลัดดา  22.0   NaN     วิทยาศาสตร์  3.45       168.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
11         NaN                     16.0      2023-12-26  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24  
17        50.0                     27.0      2024-01-08  
19        64.0                     23.0      2024-01-06

ต่อมาเราจะเห็นว่าข้อมูลในส่วนของ "GPA" นั้น
มีบางส่วนที่ติดลบและ,uบางส่วนที่มีค่ามากกว่า 4.0
เราจะแก้ไขมันด้วยการทำให้ค่าไหนที่ติดลบจะกลายเป็น NaN และค่าไหนที่มีค่ามากกว่า 4.0 ก็จะทำให้มีค่าเท่ากับ 4.0


แปลงค่า GPA ที่น้อยกว่า 0 ให้เป็น NaN

In [381]:
df_master.loc[df_master['GPA'] < 0, 'GPA'] = np.nan

แปลงค่า GPA ที่มากกว่า 4.0 ให้เป็น 4.0

In [382]:
df_master.loc[df_master['GPA'] > 4, 'GPA'] = 4.0

ต่อมาเราจะคำนวณค่า GPA เฉลี่ย (ไม่รวม NaN)


In [383]:
mean_gpa = df_master['GPA'].mean()

เติมค่าที่หาย (NaN) ในคอลัมน์ "GPA" ด้วย ค่าเฉลี่ย mean_gpa ที่คำนวณไว้ก่อนหน้า
แล้ว อัปเดตข้อมูลใน df_master

In [384]:
df_master['GPA'].fillna(mean_gpa, inplace=True)

เรียกดูข้อมูลที่แก้ไขในปัจจุบัน

In [385]:
df_master

รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ       GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.210000       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.450000       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์  3.454286       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.890000       160.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  4.000000       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.980000       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.340000       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.560000       167.0   
11         64012       นภา  22.0   ชาย     วิทยาศาสตร์  3.454286       171.0   
12         64013     บุญมี  20.0  หญิง  วิศวกรรมศาสตร์  3.230000       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.450000       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยาศาสตร์  3.670000       162.0   
15         64016      ภูมิ  19.0   ชาย     วิทยาศาสตร์  3.890000       173.0   
16         64017      มานี  20.0  หญิง  วิศวกรรมศาสตร์  3.010000       166.0   
17         64018      ยุพา  21.0  หญิง             NaN  3.230000       161.0   
19         64020     ลัดดา  22.0   NaN     วิทยาศาสตร์  3.450000       168.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
11         NaN                     16.0      2023-12-26  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24  
17        50.0                     27.0      2024-01-08  
19        64.0                     23.0      2024-01-06

ต่อมาจะเห็นว่าในส่วนของคอลัมน์ "น้ำหนัก_kg" เราจะเห็นว่าคอลัมน์นี้ยังมีส่วนที่มีค่า NaN ด้วย ซึ่งวิธีวิธีการเติมค่าด้วยค่าเฉลี่ยนั้นค่อนข้างเหมาะมากกับการแก้ไขข้อมูลในคอลัมน์นี้

ก่อนอื่นเราจะทำการคำนวณค่าเฉลี่ยของน้ำหนัก

In [386]:
mean_weight = df_master['น้ำหนัก_kg'].mean()

ต่อมาเราจะทำการเติมค่าที่หาย (NaN) ด้วยค่าเฉลี่ย

In [387]:
df_master['น้ำหนัก_kg'].fillna(mean_weight, inplace=True)

เรียกดูข้อมูลหลังจากทำการแก้ไข


In [388]:
df_master


รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ       GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.210000       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.450000       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์  3.454286       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.890000       160.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  4.000000       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.980000       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.340000       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.560000       167.0   
11         64012       นภา  22.0   ชาย     วิทยาศาสตร์  3.454286       171.0   
12         64013     บุญมี  20.0  หญิง  วิศวกรรมศาสตร์  3.230000       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.450000       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยาศาสตร์  3.670000       162.0   
15         64016      ภูมิ  19.0   ชาย     วิทยาศาสตร์  3.890000       173.0   
16         64017      มานี  20.0  หญิง  วิศวกรรมศาสตร์  3.010000       166.0   
17         64018      ยุพา  21.0  หญิง             NaN  3.230000       161.0   
19         64020     ลัดดา  22.0   NaN     วิทยาศาสตร์  3.450000       168.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0    65.000000                     15.0      2024-01-23  
1    58.000000                     20.0      2024-02-05  
3    70.000000                     10.0      2023-12-30  
4    52.000000                     30.0      2024-01-27  
6    48.000000                     35.0      2024-01-05  
7    82.000000                     18.0      2024-01-23  
8    55.000000                     22.0      2024-02-11  
9    60.000000                     28.0      2024-01-03  
11   61.733333                     16.0      2023-12-26  
12   72.000000                     24.0      2023-12-26  
13   54.000000                     19.0      2024-01-08  
14   56.000000                     26.0      2024-02-06  
15   78.000000                     14.0      2024-01-20  
16   62.000000                     21.0      2024-01-24  
17   50.000000                     27.0      2024-01-08  
19   64.000000                     23.0      2024-01-06

สรุปการทำงาน

1. โหลดและเตรียมข้อมูล
ใช้ Pandas อ่านข้อมูลจากไฟล์ student.csv และเก็บในตัวแปร df_master,
กำหนดค่าการสุ่ม (random seed) เพื่อให้ได้ผลลัพธ์ซ้ำได้
, สร้างคอลัมน์ "วันที่ลงทะเบียน" แบบสุ่มภายในช่วง ±30 วันจากวันที่ 15 มกราคม 2024, ตรวจสอบค่าที่หาย (NaN) และค่าผิดปกติในแต่ละคอลัมน์



2. ตรวจสอบและสรุปปัญหาข้อมูล
ตรวจสอบว่ามีอายุที่มากกว่า 100 ปีหรือไม่, ตรวจสอบว่า GPA น้อยกว่า 0 หรือมากกว่า 4 หรือไม่, นับจำนวนรูปแบบการเขียนชื่อคณะ (คณะ) และเพศ (เพศ), สรุปปัญหาที่พบ เช่น จำนวนค่าหาย, อายุผิดปกติ, GPA ผิดปกติ

3. เก็บสำเนาข้อมูลต้นฉบับ
สร้างตัวแปร df_original เพื่อเก็บสำเนาข้อมูลก่อนปรับปรุง เผื่อใช้เปรียบเทียบในภายหลัง

4. จัดการค่าที่หาย
เติมค่า "อายุ" ที่หายด้วยค่าเฉลี่ยของอายุ, เติมค่า "น้ำหนัก_kg" ที่หายด้วยค่าเฉลี่ยของน้ำหนัก

5. มาตรฐานการเขียนข้อความ
แปลงรูปแบบชื่อคณะ (คณะ) ให้เป็นมาตรฐานเดียวกัน เช่น "วิทย์", "Science" → "วิทยาศาสตร์", แปลงรูปแบบเพศ (เพศ) ให้เป็น "ชาย" หรือ "หญิง"



6. ตรวจจับและลบ Outlier (IQR Method)
คำนวณ Q1, Q3 และ IQR ของคอลัมน์ "อายุ", กำหนดช่วงค่าที่ถือว่าไม่เป็น outlier (lower_bound และ upper_bound), ลบแถวที่ "อายุ" อยู่นอกช่วงนี้ออกจาก df_master
(นี่คือเหตุผลที่ทำให้แถว 2, 5, 10 และ 18 หาย)

7. แก้ไขค่าผิดปกติใน GPA
ถ้า GPA < 0 → แทนด้วย NaN, ถ้า GPA > 4 → แทนด้วย 4.0, เติมค่า GPA ที่ว่างด้วยค่าเฉลี่ยของ GPA



8. แสดงข้อมูลหลังทำความสะอาด (data cleaning)
ใช้คำสั่ง print และการแสดง DataFrame เพื่อตรวจสอบข้อมูลที่ได้หลังจากทำความสะอาดเสร็จ